HealthVault/Classes/Caching/MHVThingCacheDatabase.m

```
@@ -0,0 +1,570 @@
+         record.lastSyncDate = [NSDate date];
+         record.isValid = YES;
+         
+         NSError *error = [self saveContext];
+         
+         MHVLOG(@"ThingCacheDatabase: Cache record updated %li things", things.count);
+         
+         if (error)
+         {
+             if (completion)
+             {
+                 completion(0, error);
+             }
+         }
+         else
+         {
+             if (completion)
+             {
+                 completion(things.count, nil);
+             }
+         }
+     }];
+}
+
+- (void)fetchCachedRecords:(void(^)(NSArray<id<MHVCachedRecord>> *_Nullable records))completion
+{
+    MHVASSERT_PARAMETER(completion);
+    
+    if (self.deleted)
+    {
+        if (completion)
+        {
+            completion(nil);
+        }
+        return;
+    }
+    
+    [self.managedObjectContext performBlock:^
+     {
+         NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"MHVCachedRecord"];
+         
+         NSError *error = nil;
+         NSArray *results = [self.managedObjectContext executeFetchRequest:request error:&error];
+         if (!results && error)
+         {
+             MHVLOG(@"ThingCacheDatabase: Error fetching objects: %@\n%@", [error localizedDescription], [error userInfo]);
+         }
+         
+         if (completion)
+         {
+             completion(results);
+         }
+     }];
+}
+
+- (MHVCachedRecord *)fetchCachedRecord:(NSString *)recordId
+{
+    __block MHVCachedRecord *record;
+    
+    if (self.deleted)
+    {
+        return nil;
+    }
+    
+    [self.managedObjectContext performBlockAndWait:^
+     {
+         NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"MHVCachedRecord"];
+         request.predicate = [NSPredicate predicateWithFormat:@"recordId == %@", recordId];
+         
+         NSError *error = nil;
+         NSArray *results = [self.managedObjectContext executeFetchRequest:request error:&error];
```

Here too. We need to report these errors back to the caller for traceability and program flow. The caller of this method has no idea if there was an error, or if the is just no record. prepareCacheForRecords: calls this and creates a new record if it is nil - That's not the appropriate behavior if there is an error fetching.